Creating Reals set as an alternative to S.Reals

arihantparsoya opened the pull request:
Fix for #11378.
- Created `Reals = S.Reals`
#### Sample Program:

``` python
>>> from sympy import Interval, Reals
>>> Interval(0, 1).complement(Reals)
(-oo, 0) U (1, oo)
```

arihantparsoya committed:
Creating Reals set as an alternative to S.Reals

Fix for #11378.

Created Reals set equals to S.Reals to avoid using S.Reals outside
library code.

Imported S in sets/__init__.py file instead of core/__init__.py file
since  Reals object is defined in sets.

kshitij10496 commented:
@parsoyaarihant Can you add `Complexes` as well ?

arihantparsoya committed:
Creates Complexes set as an alternative to S.Complexes

Created Reals set equals to S.Reals to avoid using S.Reals outside
library code.

kshitij10496 commented on sympy/solvers/__init__.py:
```diff
@@ -32,3 +32,7 @@
 from .solveset import solveset, linsolve, linear_eq_to_matrix
+
+from ..core.singleton import S
+Complexes = S.Complexes
```
I think this should go along with `Reals` in `sets/__init__.py`.
Is there any particular you are exposing `Complexes` here ?

parsoyaarihant replied:
yes, the there were some error : 
``` python
>>> from sympy import *
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/parsoyaarihant/sympy/sympy/__init__.py", line 49, in <module>
    from .polys import *
  File "/Users/parsoyaarihant/sympy/sympy/polys/__init__.py", line 5, in <module>
    from . import polytools
  File "/Users/parsoyaarihant/sympy/sympy/polys/polytools.py", line 54, in <module>
    from sympy.polys.domains import FF, QQ, ZZ
  File "/Users/parsoyaarihant/sympy/sympy/polys/domains/__init__.py", line 5, in <module>
    from . import domain
ImportError: cannot import name 'domain'
```

kshitij10496 replied:
> yes, the there were some errors

I would recommend you to fix this error rather than figuring a work around.

asmeurer replied:
It may be a mess to fix some of these issues. If you can't figure them out, just add imports for whatever fancy sets you can.

parsoyaarihant replied:
Should I remove `Complexes` from `solvers/__init__.py` then?

parsoyaarihant commented:
@kshitij10496 , I added for `Complexes` as well

```python
>>> from sympy import *
>>> Complexes
S.Complexes
```

kshitij10496 commented:
I made the necessary changes in `sets/__init__.py`
`Complexes = S.Complexes`

I got an `ImportError` while importing `Complexes` but with this traceback.

```python
In []: from sympy import Complexes
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-d17590a0fc31> in <module>()
----> 1 from sympy import Complexes

/home/kshitij10496/GSoC/sympy/sympy/__init__.py in <module>()
     47 from .logic import *
     48 from .assumptions import *
---> 49 from .polys import *
     50 from .series import *
     51 from .functions import *

/home/kshitij10496/GSoC/sympy/sympy/polys/__init__.py in <module>()
      3 __all__ = []
      4 
----> 5 from . import polytools
      6 __all__.extend(polytools.__all__)
      7 from .polytools import *

/home/kshitij10496/GSoC/sympy/sympy/polys/polytools.py in <module>()
     52 from mpmath.libmp.libhyper import NoConvergence
     53 
---> 54 from sympy.polys.domains import FF, QQ, ZZ
     55 from sympy.polys.constructor import construct_domain
     56 

/home/kshitij10496/GSoC/sympy/sympy/polys/domains/__init__.py in <module>()
      7 from .domain import *
      8 
----> 9 from . import finitefield
     10 __all__.extend(finitefield.__all__)
     11 from .finitefield import *

/home/kshitij10496/GSoC/sympy/sympy/polys/domains/finitefield.py in <module>()
      5 from sympy.polys.domains.field import Field
      6 from sympy.polys.domains.simpledomain import SimpleDomain
----> 7 from sympy.polys.domains.groundtypes import SymPyInteger
      8 from sympy.polys.domains.modularinteger import ModularIntegerFactory
      9 

/home/kshitij10496/GSoC/sympy/sympy/polys/domains/groundtypes.py in <module>()
     11 PythonComplex = builtins.complex
     12 
---> 13 from .pythonrational import PythonRational
     14 
     15 from sympy.core.numbers import (

/home/kshitij10496/GSoC/sympy/sympy/polys/domains/pythonrational.py in <module>()
     11 from sympy.core.sympify import converter
     12 from sympy.core.numbers import Rational
---> 13 from sympy.printing.defaults import DefaultPrinting
     14 from sympy.utilities import public
     15 

/home/kshitij10496/GSoC/sympy/sympy/printing/__init__.py in <module>()
      7 from .python import python, print_python
      8 from .ccode import ccode, print_ccode
----> 9 from .fcode import fcode, print_fcode
     10 from .jscode import jscode, print_jscode
     11 from .julia import julia_code

/home/kshitij10496/GSoC/sympy/sympy/printing/fcode.py in <module>()
     25 from sympy.core.compatibility import string_types, range
     26 from sympy.core.function import Function
---> 27 from sympy.sets import Range
     28 from sympy.codegen.ast import Assignment
     29 from sympy.printing.codeprinter import CodePrinter

/home/kshitij10496/GSoC/sympy/sympy/sets/__init__.py in <module>()
      8 
      9 Reals = S.Reals
---> 10 Complexes = S.Complexes

/home/kshitij10496/GSoC/sympy/sympy/core/singleton.py in __getattr__(self, name)
     45                 name, self))
     46         class_to_install = self._classes_to_install[name]
---> 47         value_to_install = class_to_install()
     48         self.__setattr__(name, value_to_install)
     49         del self._classes_to_install[name]

/home/kshitij10496/GSoC/sympy/sympy/core/singleton.py in __call__(self, *args, **kwargs)
    106         if self not in Singleton._instances:
    107             Singleton._instances[self] = \
--> 108                 super(Singleton, self).__call__(*args, **kwargs)
    109                 # Invokes the standard constructor of SomeClass.
    110         return Singleton._instances[self]

/home/kshitij10496/GSoC/sympy/sympy/sets/fancysets.py in __new__(cls)
   1427 
   1428     def __new__(cls):
-> 1429         return ComplexRegion.__new__(cls, S.Reals*S.Reals)
   1430 
   1431     def __eq__(self, other):

/home/kshitij10496/GSoC/sympy/sympy/sets/fancysets.py in __new__(cls, sets, polar)
   1119 
   1120     def __new__(cls, sets, polar=False):
-> 1121         from sympy import sin, cos
   1122 
   1123         x, y, r, theta = symbols('x, y, r, theta', cls=Dummy)

ImportError: cannot import name 'sin'

```

Is this a case of circular import ?
Ping @asmeurer @aktech

parsoyaarihant commented:
@kshitij10496 , I recieve the same error too once, and when I run the `import` command again, it gives error from `Polys`. What can be done about it?

arihantparsoya committed:
Handle relative import

Upabjojr commented on sympy/sets/__init__.py:
```diff
@@ -3,3 +3,7 @@
+
+from sympy.core.singleton import S
+Reals = S.Reals
+del S
```
is this `del` necessary?

asmeurer replied:
It's a good idea to `del` names that you don't want to export in `__init__.py` files.

asmeurer commented on sympy/sets/__init__.py:
```diff
@@ -3,3 +3,7 @@
 from .contains import Contains
 from .conditionset import ConditionSet
+
+from sympy.core.singleton import S
```
Use a relative import

parsoyaarihant replied:
Using relative import gives the following error for python3 in travis build [here](https://travis-ci.org/sympy/sympy/jobs/143955456):
```python
Exception occurred:
  File "../sympy/sets/__init__.py", line 7, in <module>
    from ..core.singleton import S
ValueError: attempted relative import beyond top-level package
```

asmeurer replied:
Maybe it would be better to just add 

```py
from .fancysets import Reals

Reals = Reals()
```

so we don't import `S` at all.

asmeurer commented:
Some things you can try with the imports:

- Make them full imports (like `from sympy.functions.elementary.trigonometric import sin` instead of `from sympy import sin`)

- Move them into the `if` block that needs them

asmeurer commented:
For what it's worth, I don't think it was a good idea to singletonize the fancy sets. They don't really need to be singletonized. It may save a bit of memory, but I'm sceptical that they are really used that often, but every entry in `S` slows down the import time of SymPy, since an instance is created when the class is defined.

parsoyaarihant commented:
>Make them full imports (like from sympy.functions.elementary.trigonometric import sin instead of from sympy import sin)

I have already tried that and the library still raises issues in the build.

arihantparsoya committed:
Enabled relative import

Defined Reals without importing S for correct relative import.

Using ‘from ..core.singleton import S’
 gives the following error:
 attempted relative import beyond top-level package
